File: python/DimRed.py
'''
A feedforward neural network to reduce dimensionality

Each irreducible has its own dimensionality reduction network
'''

from typing import List
import torch
import torch.nn

import SSAIC

class DimRedNet(torch.nn.Module):
    # max_depth <= 0 means unlimited
    def __init__(self, init_dim:int, max_depth=0) -> None:
        super(DimRedNet, self).__init__()
        # Determine depth
        depth = init_dim - 1
        if max_depth > 0 and depth > max_depth: depth = max_depth
        # Fully connected layer to reduce dimensionality
        self.fc = torch.nn.ModuleList()
        for i in range(depth):
            self.fc.append(torch.nn.Linear(
                init_dim - i, init_dim - i - 1,
                bias=False))
        # Fully connected layer to inverse the reduction
        self.fc_inv = torch.nn.ModuleList()
        for i in range(depth):
            self.fc_inv.append(torch.nn.Linear(
                init_dim - i - 1, init_dim - i,
                bias=False))

    # For pretraining
    def forward(self, x:torch.Tensor) -> torch.Tensor:
        y = x.clone()
        # Reduce dimensionality
        for layer in self.fc:
            y = layer(y)
            y = torch.tanh(y)
        # Inverse the reduction
        for layer in reversed(self.fc_inv):
            y = torch.tanh(y)
            y = layer(y)
        return y

    def freeze(self, freeze:int) -> None:
        assert freeze < len(self.fc), "All layers are frozen, so nothing to train"
        for i in range(freeze):
            self.fc[i].weight.requires_grad = False
            self.fc_inv[i].weight.requires_grad = False

# The dimensionality reduction networks of each irreducible
net_list = None

def define_DimRed():
    global net_list
    net_list = []

def reduce(x:List[torch.Tensor]) -> List[torch.Tensor]:
    y = x.copy()
    for i in range(SSAIC.NIrred):
        y[i] = net_list[i].reduce(y[i])
    return y

def inverse(x:List[torch.Tensor]) -> List[torch.Tensor]:
    y = x.copy()
    for i in range(SSAIC.NIrred):
        y[i] = net_list[i].inverse(y[i])
    return y

File: cpp/source/net/DimRed.cpp
// A feedforward neural network to reduce dimensionality

#include <torch/torch.h>

#include "../../Cpp-Library_v1.0.0/TorchSupport.hpp"

#include "../../include/SSAIC.hpp"
#include "../../include/net.hpp"

namespace DimRed {

// max_depth == 0 means unlimited
Net::Net(const size_t & init_dim, const size_t & max_depth) {
    // Determine depth
    size_t depth = init_dim - 1;
    if (max_depth > 0 && depth > max_depth) depth = max_depth;
    // Fully connected layer to reduce dimensionality
    fc.resize(depth);
    for (size_t i = 0; i < depth; i++) {
        fc[i] = new torch::nn::Linear{nullptr};
        * fc[i] = register_module("fc-"+std::to_string(i),
            torch::nn::Linear(torch::nn::LinearOptions(
            init_dim - i, init_dim - i - 1)
            .bias(false)));
    }
    // Fully connected layer to inverse the reduction
    fc_inv.resize(depth);
    for (size_t i = 0; i < depth; i++) {
        fc_inv[i] = new torch::nn::Linear{nullptr};
        * fc_inv[i] = register_module("fc_inv-"+std::to_string(i),
            torch::nn::Linear(torch::nn::LinearOptions(
            init_dim - i - 1, init_dim - i)
            .bias(false)));
    }
}

// For pretraining
torch::Tensor Net::forward(const at::Tensor & x) {
    torch::Tensor y = x.clone();
    // Reduce dimensionality
    for (auto & layer : this->fc) {
        y = (*layer)->forward(y);
        y = torch::tanh(y);
    }
    // Inverse the reduction
    for (auto layer = this->fc_inv.rbegin();
        layer != this->fc_inv.rend(); ++layer) {
        y = torch::tanh(y);
        y = (**layer)->forward(y);
    }
    return y;
}

void Net::warmstart(const std::string & chk, const size_t & chk_depth) {
    //size_t init_dim = (*fc[0])->options.in_features();
    //auto warm_net = std::make_shared<Net>(init_dim, chk_depth_);
    auto warm_net = std::make_shared<Net>((*fc[0])->options.in_features(), chk_depth);
    warm_net->to(torch::kFloat64);
    torch::load(warm_net, chk);
    torch::NoGradGuard no_grad;
    for (size_t i = 0; i < (fc.size() < warm_net->fc.size() ? fc.size() : warm_net->fc.size()); i++) {
        CL::TS::copy((*fc    [i])->weight, (*(warm_net->fc    [i]))->weight);
        CL::TS::copy((*fc_inv[i])->weight, (*(warm_net->fc_inv[i]))->weight);
    }
    warm_net.reset();
}

// 
void Net::freeze(const size_t & freeze) {
    
}

} // namespace DimRed